### ISSUE TITLE
A schema with a Lambda Resolver in GraphiQL had an empty response error.

### ISSUE THREAD

**Issue by nikaera** [bboure/serverless-appsync-simulator#51]
**Describe the bug**
I tried to start it offline with serverless-appsync-simulator by writing serverless.yml as follows.

```yaml
# serverless.yml
custom:
  dynamodb:
    start:
      seed: true
      migrate: true
      inMemory: true
    stages:
      - dev
    seed:
      dev:
        sources:
          - table: ${self:service.name}_${self:provider.stage}_User
            sources: [./test/migrations/users.json]
          - table: ${self:service.name}_${self:provider.stage}_Channel
            sources: [./test/migrations/channels.json]
#...
  appSync:
    dataSources:
      - type: AWS_LAMBDA
        name: GetUserByName
        description: "Function to get user information based on a user's name"
        config:
          functionName: GetUserByName
          iamRoleStatements:
            - Effect: "Allow"
              Action:
                - "lambda:invokeFunction"
              Resource:
                - "*"
#...
    mappingTemplates:
      - dataSource: GetUserByName
        type: Query
        field: getUserByName
        request: functions.request.vtl
        response: functions.response.vtl
#...
    functionConfigurations:
      - dataSource: GetUserByName
        name: GetUserByName
        request: functions.request.vtl
        response: functions.response.vtl
#...
functions:
  GetUserByName:
    role: ServerlessRoleDev
    handler: AppSync.GetUserByName
```
```typescript
// handler.ts
// ...
interface GetUserByNameInput {
  username: string;
}
export const GetUserByName: LambdaResolver = async (
  event: GetUserByNameInput
) => {
  console.log(event);
  const username = event.username;
  const user = await entityManager.getUserByName(username);

  return { ...event, ...user };
};
// ...
```

I have verified that the `IS_OFFLINE=1 sls offline start` can be started successfully, and the `aws lambda invoke` command can be used.
```bash
aws lambda invoke /dev/null \
    --endpoint-url http://localhost:3002 \
    --function-name live-service-dev-GetUserByName \
    --payload '{ "username": "kadinche3" }' --cli-binary-format raw-in-base64-out
{
    "StatusCode": 200
}

{ username: 'tes' }
{"data":{"listUsers":{"items":[{"id":"8d7105ff-fd2a-0d77-d817-cb032e7626b3","username":"tes","email":"tes@test.com","type":"Admin","channel_ids":["eedc2c40-f714-cb81-29a6-6c8abae660c3"],"created_at":"12345","updated_at":"12345","__typename":"User"}],"nextToken":null,"__typename":"UserConnection"}},"loading":false,"networkStatus":7,"stale":false}
offline: (λ: GetUserByName) RequestId: ckga7oh4j00007ujl92u4aprq  Duration: 962.32 ms  Billed Duration: 1000 ms
```

However, when trying to execute the same command with GraphiQL, the response was not returned. **Also, when I checked the terminal, there was no output of access from GraphiQL.** (What I find most suspicious right now is that it doesn't seem to be able to run Lambda Resolver.)

<img width="1217" alt="スクリーンショット 2020-10-15 12 03 24" src="https://user-images.githubusercontent.com/1802476/96073754-0b4e2300-0ee2-11eb-9c55-d66199296032.png">

I know it looks like a timeout error as well, could you please tell what might be causing it?
Please let me know if there is any other information that might be helpful or an item I should verify. 🙏 🙇 

**To Reproduce**
I've described the procedure in "Describe the bug" so I'll describe my environment below.
- Serverless
  - Framework Core: 2.7.0 (local)
  - Plugin: 4.1.0
  - SDK: 2.3.2
  - Components: 3.2.1
- serverless-offline 6.8.0
- serverless-appsync-plugin 1.4.0
- serverless-appsync-simulator 0.7.0
- amplify-appsync-simulator 1.14.0

**Expected behavior**
I thought the result would be like `aws lambda invoke` when I used the schema with the Lambda Resolver in GraphiQL.

**Screenshots**

**Additional context**
- All of the DynamoDB resolvers worked successfully in GraphiQL as well. (ex. getUser, listUsers, deleteUser, createUser, etc.)
<img width="1088" alt="スクリーンショット 2020-10-15 13 25 07" src="https://user-images.githubusercontent.com/1802476/96077118-f2e20680-0ee9-11eb-9bc8-d994d2a2dfda.png">
<img width="1007" alt="スクリーンショット 2020-10-15 13 25 33" src="https://user-images.githubusercontent.com/1802476/96077126-fa091480-0ee9-11eb-9d60-8025912a7487.png">

- I have done sls deploy and verified in the AWS Console, and the function with Lambda Resolver has been successfully executed.
<img width="646" alt="スクリーンショット 2020-10-15 12 52 51" src="https://user-images.githubusercontent.com/1802476/96075131-5c134b00-0ee5-11eb-81b7-f184dfb356b2.png">

- I have seed data in the user table, but the number of seed data in the user table is 5, which is not a lot.

Anyway, I'm very grateful for your plugins, I use them almost every time I develop with Serverless Framework. 🛠️  I've been investigating this case by myself, but I wanted your help and created an issue. 📝 🙇

**Comment by bboure** [comment]
Hi @nikaera 
Thanks for the detailed information.
What do you see in your terminal console when you execute the request?
Any log or error?

**Comment by nikaera** [comment]
Hi @bboure 
> What do you see in your terminal console when you execute the request?

The `aws invoke` command printed a log, but GraphiQL didn't print anything to the terminal, even after the GraphiQL error.

**Comment by nikaera** [comment]
I added a console.log below to see if I was trying to run the Lambda function when I made a request from GraphiQL.
https://github.com/dherault/serverless-offline/blob/c4688c60cab1d029f1c794d9f3fdaf5a9edf8a3c/src/lambda/LambdaFunction.js#L192

The result was that running `aws invoke` printed out the console.log, but making a GraphiQL request did not... 🤔

**Comment by nikaera** [comment]
Hi @bboure, I may have found the cause.

I first checked the following template project to see if it worked.
https://github.com/daisuke-awaji/serverless-appsync-offline-typescript-template

I have confirmed that the query with the Lambda resolver is working properly.
I then checked the differences between the project and the project where the Lambda resolver was not working properly.

I'm using the [serverless-functions-base-path](https://github.com/kevinrambaud/serverless-functions-base-path/) plug-in for the Serverless Framework. Using serverless-functions-base-path, you can change the file path where the handler is placed.
I wondered if this was the reason why I hadn't been able to discover the Lambda resolver function.

So I experimented with a project that I was able to validate.
I tried the following Lambda resolver query.
https://github.com/daisuke-awaji/serverless-appsync-offline-typescript-template/blob/master/serverless.yml#L47-L57
https://github.com/daisuke-awaji/serverless-appsync-offline-typescript-template/blob/master/serverless.yml#L60-L64

After running it without any modifications, I was able to successfully run it in GraphiQL.
<img width="1106" alt="スクリーンショット 2020-10-16 12 21 10" src="https://user-images.githubusercontent.com/1802476/96209252-1cfafd80-0faa-11eb-84f4-d1095dede3b5.png">

Next, I experimented with changing the handler path with serverless-functions-base-path.
I have placed the handler.ts file in the lib folder. (lib/handler.ts)
```yaml
#serverless.yml
service:
  name: myService
# app and org for use with dashboard.serverless.com
#app: your-app-name
#org: your-org-name

plugins:
  - serverless-functions-base-path
  - serverless-webpack
  - serverless-appsync-plugin
  - serverless-dynamodb-local
  - serverless-appsync-simulator
  - serverless-offline

custom:
  functionsBasePath: lib
  appsync-simulator:
    location: ".webpack/service/lib"
#...
```

I tested again with `sls offline start` and found the same problem as with Issue.
<img width="1064" alt="スクリーンショット 2020-10-16 12 26 15" src="https://user-images.githubusercontent.com/1802476/96209586-effb1a80-0faa-11eb-8bb2-4499c9834b37.png">

However, I haven't tested whether the problem can be solved without using serverless-functions-base-path in my current project, so I'm going to check now if it really works. 📝

**Comment by nikaera** [comment]
GraphiQL is now able to run, but only the first request succeeds, and the second and subsequent requests produce the following error...
<img width="1088" alt="スクリーンショット 2020-10-16 14 44 43" src="https://user-images.githubusercontent.com/1802476/96217768-41f96b80-0fbe-11eb-9d5d-829612f874da.png">
```bash
#...
Lambda Data source failed with the following error
Command failed with exit code 1: /Users/nika/.anyenv/envs/nodenv/versions/12.18.3/bin/node /Users/nika/Documents/workspace/LiveINN-AWS-Serverless/node_modules/amplify-nodejs-function-runtime-provider/lib/utils/execute.js

{"result":{"name":"myapp","version":"0.0.1"},"error":null}

Lambda Data source failed with the following error
Command failed with exit code 1: /Users/nika/.anyenv/envs/nodenv/versions/12.18.3/bin/node /Users/nika/Documents/workspace/LiveINN-AWS-Serverless/node_modules/amplify-nodejs-function-runtime-provider/lib/utils/execute.js

{"result":{"name":"myapp","version":"0.0.1"},"error":null}
#...
```

**Comment by bboure** [comment]
Thanks for all the details.
Looks like the other issue your are talking about has been fixed (https://github.com/aws-amplify/amplify-cli/pull/5385)

However, I think that part of the code is deprecated and we are not using it anymore in this project.
What version of serverless-appsync-simulator are you using?

**Comment by bboure** [comment]
Never mind, I already see you are using a recent version `amplify-nodejs-function-runtime-provider/lib/utils/execute.js`  is the right path.
I wonder if the same fix should be applied [here](https://github.com/aws-amplify/amplify-cli/blob/master/packages/amplify-nodejs-function-runtime-provider/src/utils/execute.ts#L108)

**Comment by nikaera** [comment]
Thank you for your confirmation. 🙏 

> What version of serverless-appsync-simulator are you using?

I'm using 0.7.0 right now. 🙋 

```fish
⊨ cat yarn.lock | grep serverless-appsync-simulator
serverless-appsync-simulator@^0.7.0:
  resolved "https://registry.yarnpkg.com/serverless-appsync-simulator/-/serverless-appsync-simulator-0.7.0.tgz#d3aa7f5e205290fbac409bd39d14a541f0747459"
```

> I wonder if the same fix should be applied here

I am also curious about this. I have looked at the logs and the console shows an error, but when I look at the results of the run it appears to be successful. ⬇️ 🤔 
```fish
#...
offline: [HTTP] server ready: http://localhost:3000 🚀
offline: 
offline: Enter "rp" to replay the last request
Lambda Data source failed with the following error
Command failed with exit code 1: /Users/nika/.anyenv/envs/nodenv/versions/12.18.3/bin/node /Users/nika/Documents/workspace/LiveINN-AWS-Serverless/node_modules/amplify-nodejs-function-runtime-provider/lib/utils/execute.js
{ username: 'test' }
{"data":{"listUsers":{"items":[{"id":"8d7105ff-fd2a-0d77-d817-cb032e7626b0","username":"test","email":"test@test.com","type":"Admin","channel_ids":["eedc2c40-f714-cb81-29a6-6c8abae660c3"],"created_at":"12345","updated_at":"12345","__typename":"User"}],"nextToken":null,"__typename":"UserConnection"}},"loading":false,"networkStatus":7,"stale":false}

{"result":{"username":"test","id":"8d7105ff-fd2a-0d77-d817-cb032e7626b0","email":"test@test.com","type":"Admin","channel_ids":["eedc2c40-f714-cb81-29a6-6c8abae660c3"],"created_at":"12345","updated_at":"12345","__typename":"User"},"error":null}
{ username: 'test' }
{"data":{"listUsers":{"items":[{"id":"8d7105ff-fd2a-0d77-d817-cb032e7626b0","username":"test","email":"test@test.com","type":"Admin","channel_ids":["eedc2c40-f714-cb81-29a6-6c8abae660c3"],"created_at":"12345","updated_at":"12345","__typename":"User"}],"nextToken":null,"__typename":"UserConnection"}},"loading":false,"networkStatus":7,"stale":false}
#...
```

<img width="1090" alt="スクリーンショット 2020-10-16 20 58 34" src="https://user-images.githubusercontent.com/1802476/96255741-6111f080-0ff2-11eb-9a63-4e592e58add7.png">

When I hit the button repeatedly without error, the following log will not be printed, but the resulting log will be the same.
```
# Output in case of an error.
Lambda Data source failed with the following error
Command failed with exit code 1: /Users/nika/.anyenv/envs/nodenv/versions/12.18.3/bin/node /Users/nika/Documents/workspace/LiveINN-AWS-Serverless/node_modules/amplify-nodejs-function-runtime-provider/lib/utils/execute.js

{ username: 'test' }
{"data":{"listUsers":{"items":[{"id":"8d7105ff-fd2a-0d77-d817-cb032e7626b0","username":"test","email":"test@test.com","type":"Admin","channel_ids":["eedc2c40-f714-cb81-29a6-6c8abae660c3"],"created_at":"12345","updated_at":"12345","__typename":"User"}],"nextToken":null,"__typename":"UserConnection"}},"loading":false,"networkStatus":7,"stale":false}

{"result":{"username":"test","id":"8d7105ff-fd2a-0d77-d817-cb032e7626b0","email":"test@test.com","type":"Admin","channel_ids":["eedc2c40-f714-cb81-29a6-6c8abae660c3"],"created_at":"12345","updated_at":"12345","__typename":"User"},"error":null}
{ username: 'test' }
{"data":{"listUsers":{"items":[{"id":"8d7105ff-fd2a-0d77-d817-cb032e7626b0","username":"test","email":"test@test.com","type":"Admin","channel_ids":["eedc2c40-f714-cb81-29a6-6c8abae660c3"],"created_at":"12345","updated_at":"12345","__typename":"User"}],"nextToken":null,"__typename":"UserConnection"}},"loading":false,"networkStatus":7,"stale":false}

# Output on success.
{ username: 'test' }
{"data":{"listUsers":{"items":[{"id":"8d7105ff-fd2a-0d77-d817-cb032e7626b0","username":"test","email":"test@test.com","type":"Admin","channel_ids":["eedc2c40-f714-cb81-29a6-6c8abae660c3"],"created_at":"12345","updated_at":"12345","__typename":"User"}],"nextToken":null,"__typename":"UserConnection"}},"loading":false,"networkStatus":7,"stale":false}
```

What I'm wondering is why the log output looks like it's running Lambda twice in a row on an error, and why the output includes a JSON log with a key named result.

**Comment by nikaera** [comment]
@bboure Oh, I just thought I'd verify, based on your opinion, by changing exit(1) to exit(0) and it always succeeds in running Lambda!! 👀 
https://github.com/aws-amplify/amplify-cli/blob/master/packages/amplify-nodejs-function-runtime-provider/src/utils/execute.ts#L108

**Comment by bboure** [comment]
great, 
If this fixes your issue, a PR should be open on aws-amplify

**Comment by somq** [comment]
> great,
> If this fixes your issue, a PR should be open on aws-amplify

I confirm your fix @bboure ,
There is already a PR merged, prolly release in `amplify-appsync-simulator@1.23..2` (didnt verify tho)

- Related issue: https://github.com/aws-amplify/amplify-cli/issues/5378
- PR: https://github.com/aws-amplify/amplify-cli/pull/5385/files

A `amplify-appsync-simulator` upgrade should not hurt anyway

**Comment by nikaera** [comment]
@bboure Thank you for the suggestion. I have created a PR. 🛠️ 
https://github.com/aws-amplify/amplify-cli/pull/5608/files

@somq Thank you for your comment. 🙏
I fixed `amplify-nodejs-function-runtime provider` instead of `amplify-util-mock`, so I've created a new PR. 🙋

**Comment by BracketJohn** [comment]
We're also facing this issue when chaining lambda functions as query results. If we chain two functions `f1` and `f2` for a `Query`, we only get a valid return iff `f1` and `f2` return in time. Else the exit code `1` of either function stops GQL from resolving correctly.

Very much looking forward to the fix provided in https://github.com/aws-amplify/amplify-cli/pull/5608!

**Comment by somq** [comment]
> 
> 
> We're also facing this issue when chaining lambda functions as query results. If we chain two functions `f1` and `f2` for a `Query`, we only get a valid return iff `f1` and `f2` return in time. Else the exit code `1` of either function stops GQL from resolving correctly.
> 
> Very much looking forward to the fix provided in [aws-amplify/amplify-cli#5608](https://github.com/aws-amplify/amplify-cli/pull/5608)!

@BracketJohn I monkeypatched the issue in the meantime on my side

package.json
```json
{
"monkeypatch-appsync-simulator": "cp _sls-monkeypatch-appsync-simulator/execute.js node_modules/amplify-nodejs-function-runtime-provider/lib/utils/execute.js",
"monkeypatch-appsync-simulator:issue": "https://github.com/aws-amplify/amplify-cli/pull/5608",
"postinstall": "npm run monkeypatch-appsync-simulator"
}
````

**Comment by nikaera** [comment]
The following PRs have been successfully merged and this issue is now closed! 🎉 
https://github.com/aws-amplify/amplify-cli/pull/5608/files

It was very emotional to be able to commit to OSS for the first time in my life.
I would like to thank everyone involved in this issue. Thank you so much guys! 🙌
